oio/container/sharding.py

```diff
@@ -0,0 +1,213 @@
+            raise ValueError('Expected an "index" in the shard range')
+        try:
+            shard_index = int(shard_index)
+        except ValueError:
+            raise ValueError('Expected a number for the "index"')
+        if shard_index < 0:
+            raise ValueError('Expected a positive number for the "index"')
+        new_shard['index'] = shard_index
+
+        shard_lower = shard.get('lower')
+        if shard_lower is None:
+            raise ValueError('Expected a "lower" in the shard range')
+        if six.PY3 and isinstance(shard_lower, six.binary_type):
+            shard_lower = shard_lower.decode('utf-8')
+        elif six.PY2 and isinstance(shard_lower, six.text_type):
+            shard_lower = shard_lower.encode('utf-8')
+        elif not isinstance(shard_lower, str):
+            raise ValueError('Expected a string for the "lower"')
+        new_shard['lower'] = shard_lower
+
+        shard_upper = shard.get('upper')
+        if shard_upper is None:
+            raise ValueError('Expected a "upper" in the shard range')
+        if six.PY3 and isinstance(shard_upper, six.binary_type):
+            shard_upper = shard_upper.decode('utf-8')
+        elif six.PY2 and isinstance(shard_upper, six.text_type):
+            shard_upper = shard_upper.encode('utf-8')
+        elif not isinstance(shard_upper, str):
+            raise ValueError('Expected a string for the "upper"')
+        new_shard['upper'] = shard_upper
+
+        if shard_lower != '' and shard_upper != '' \
+                and shard_lower >= shard_upper:
+            raise ValueError('Expected a "upper" greater the "lower"')
+
+        if timestamp:
+            shard_timestamp = shard.get('timestamp')
+            if shard_upper is None:
+                raise ValueError('Expected a "timestamp" in the shard range')
+            try:
+                shard_timestamp = int(shard_timestamp)
+            except ValueError:
+                raise ValueError('Expected microseconds for the "timestamp"')
+            if shard_timestamp < 0:
+                raise ValueError(
+                    'Expected positive microseconds for the "timestamp"')
+            new_shard['timestamp'] = shard_timestamp
+
+        return new_shard
+
+    def _format_shards(self, shards):
+        if not isinstance(shards, list):
+            raise ValueError('Expected a list of shard ranges')
+
+        if len(shards) < 2:
+            raise ValueError('Expected at least 2 shards')
+
+        new_shards = list()
+        for shard in shards:
+            new_shards.append(self._format_shard(shard))
+
+        new_shards.sort(key=lambda new_shard: new_shard['index'])
+        first_shard_index = new_shards[0]['index']
+        last_shard_index = new_shards[-1]['index']
+        previous_shard_upper = ''
+        for i, new_shard in enumerate(new_shards):
+            if new_shard['index'] != i:
+                raise ValueError('Missing "index" %d' % i)
```

Comment:
IMO error message is wrong, it should be something like "Invalid `"index" %d, found %d" % (new_shard['index'], i)`

Reply:
The shards are sorted by index. So if the index is different than expected, this index is missing.